### PR TITLE
Add mount detection

### PR DESCRIPTION
**Pull request by jgrund**
use findmnt --poll in a small daemon that emits on changes.

This will collect then emit a command to the scanner daemon
where we can add it under a mounts key to the reducer.

Signed-off-by: Joe Grund <joe.grund@intel.com>
Signed-off-by: Tom Nabarro <tom.nabarro@intel.com>

### CODE REVIEW

File: IML.DeviceScannerDaemon/src/Mount.fs [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+let create target source fstype options =
+  {
+    target = target;
+    source = source;
+    fstype = fstype;
+    options = options
+  }
+
+type LocalMounts = Map<Mount.MountPoint, Data>
```
Comment (jgrund) [inline IML.DeviceScannerDaemon/src/Mount.fs]: Any advantage to this being a `Map` over a `Set`?

File: IML.Types/src/CommandTypes.fs [hunk 2]
```diff
@@ -6,6 +6,13 @@ module IML.Types.CommandTypes
 
 open Fable.Core
 
+[<Erase>]
+type BdevPath = BdevPath of string
+[<Erase>]
+type MountPoint = MountPoint of string
+[<Erase>]
+type FsType = FsType of string
```
Comment (jgrund) [inline IML.Types/src/CommandTypes.fs]: Dups of module below.

File: IML.DeviceScannerDaemon/src/Mount.fs [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+  /// mounted block device
+  source: Mount.BdevPath; //Mount.Bdev;
+  /// filesystem type
+  fstype: Mount.FsType;
+  /// mount options
+  options: Mount.Options;
+}
+
+let create target source fstype options =
+  {
```
Comment (jgrund) [inline IML.DeviceScannerDaemon/src/Mount.fs]: You should move this type into `CommandTypes` and then send that over the wire. If you do that you don't need to create the type on this side.

Of course, that means the Umount variant needs to send the whole record, but I don't think that's an issue.

File: IML.DeviceScannerDaemon/src/Mount.fs (hunk 1 above)
Comment (tanabarr) [reply IML.DeviceScannerDaemon/src/Mount.fs]: because we can remove by just selecting mount point as key

File: IML.DeviceScannerDaemon/src/Mount.fs [hunk 4]
```diff
@@ -5,38 +5,17 @@
...
 let update (localMounts:LocalMounts) (x:MountCommand):Result<LocalMounts, exn> =
   match x with
-    | Mount (target, source, fstype, options)
-    | Remount (target, source, fstype, options)
-    | Move (target, source, fstype, options) ->
-      Map.add target (create target source fstype options) localMounts
+    | Mount y
+    | Remount y
+    | Move y ->
+      Set.add y localMounts
```
Comment (jgrund) [inline IML.DeviceScannerDaemon/src/Mount.fs]: Does a remove come before a move?

I.E. if you get a move, do you need to remove the old mount first here?

File: IML.Listeners/MountEmitter/src/Main.fs [hunk 5]
```diff
@@ -26,30 +26,36 @@ let transform (x:Stream.Readable<string>) =
...
+            |> Mount |> MountCommand |> Ok
+        | "remount" ->
+          m
+            |> Remount |> MountCommand |> Ok
+        | "move" ->
+          m
+            |> Move |> MountCommand |> Ok
+        | "unmount" ->
+          m
+            |> Umount |> MountCommand |> Ok
```
Comment (jgrund) [inline IML.Listeners/MountEmitter/src/Main.fs]: Why are you wrapping these cases in a result?

File: IML.DeviceScannerDaemon/src/Mount.fs (hunk 4 above)
Comment (tanabarr) [reply IML.DeviceScannerDaemon/src/Mount.fs]: have to think about this one a bit more in terms of how to find out what has moved in order to remove it

Comment (tanabarr) [reply IML.DeviceScannerDaemon/src/Mount.fs]: I suppose I need to remove the Mount that was pointing to the source path first and then add the Mount

File: IML.Listeners/MountEmitter/src/Transform.fs [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+          m
+            |> Remount |> MountCommand |> Ok
+        | "move" ->
+          m
+            |> Move |> MountCommand |> Ok
+        | "umount" ->
+          m
+            |> Umount |> MountCommand |> Ok
+        | _ ->
+          failwithf "did not get expected row, got %A" a
```
Comment (jgrund) [inline IML.Listeners/MountEmitter/src/Transform.fs]: Instead of `failwithf`, use `Error (exn (sprintf "did not get expected row, got %A" a))`.

This call is expecting back a `Result` so we should return in the error case instead of raising.

File: IML.DeviceScannerDaemon/test/MountTest.fs [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+
+  matcher addMount
+
+  matcher unMount
+
+  matcher reMount
+
+  matcher moveMount
+
+test "Events on existing mount" <| fun () ->
```
Comment (jgrund) [inline IML.DeviceScannerDaemon/test/MountTest.fs]: Most of the things above here don't look related to mounts.

File: IML.DeviceScannerDaemon/src/Mount.fs [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+open MountTypes
+
+
+let update (localMounts:LocalMounts) (x:MountCommand):Result<LocalMounts, exn> =
+  match x with
+    // fixme: remove old entry matching "source", add new entry
+    // | AddMount _
+    // ???
+    // | ReplaceMount _ ->
+      // localMounts
```
Comment (jgrund) [inline IML.DeviceScannerDaemon/src/Mount.fs]: Note for impl.

File: IML.DeviceScannerDaemon/test/MountTest.fs [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+open IML.CommonLibrary
+open IML.Types.CommandTypes
+open Fable.PowerPack
+open Thot.Json
+open IML.Types
+
+let matcher localMounts x =
+  x
+    |> update localMounts
+    // |> Result.map Set.toList
```
Comment (jgrund) [inline IML.DeviceScannerDaemon/test/MountTest.fs]: Dead code here.

File: IML.DeviceScannerDaemon/src/Handlers.fs [hunk 10]
```diff
@@ -4,8 +4,10 @@
 
 module IML.DeviceScannerDaemon.Handlers
 
+open Mount
```
Comment (jgrund) [inline IML.DeviceScannerDaemon/src/Handlers.fs]: Unused import.

File: IML.Listeners/MountEmitter/src/Transform.fs [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+
+let private toRow = function
+  | [| a; b; c; d; e; |] -> Ok (a, b, c, d, e)
+  | [| a; b; c; d; |] -> Ok ("mount", a, b, c, d)
+  | x -> failwithf "did not get expected row data, got %A" x
+
+let private pollHeader = ("ACTION", "TARGET", "SOURCE", "FSTYPE", "OPTIONS")
+let private notPollHeader x = x <> pollHeader
+let private listHeader = ("mount", "TARGET", "SOURCE", "FSTYPE", "OPTIONS")
+let private notListHeader x = x <> listHeader
```
Comment (johnsonw) [inline IML.Listeners/MountEmitter/src/Transform.fs]: It feels odd to have a function check if something is not something else. I think it would be cleaner to do something like this:
```
let private isPollHeader x = x = pollHeader
```

Then below:
```
|> Stream.filter (isPollHeader >> not >> Ok)
```

File: IML.DeviceScannerDaemon/systemd-units/device-scanner.service [hunk 12]
```diff
@@ -2,6 +2,8 @@
 Description=IML Device Scanner Daemon
 RefuseManualStart=true
 DefaultDependencies=false
+Requires=mount-emitter.service
```
Comment (jgrund) [inline IML.DeviceScannerDaemon/systemd-units/device-scanner.service]: I don't think we need an explict `Requires` / `After` here. 

The mount-emitter will write to the device-scanner socket, which will explicitly start the scanner if it hasn't already.

Comment (tanabarr) [reply IML.DeviceScannerDaemon/systemd-units/device-scanner.service]: requires ensures that the service is started without having to do it explicitly, after makes sure that the scanner is only started after Mount-emitter

File: IML.Listeners/MountEmitter/src/Transform.fs [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+    |> Stream.map toRow
+    |> Stream.filter (notPollHeader >> Ok)
+    |> Stream.filter (notListHeader >> Ok)
+    |> Stream.map(function
+      | a, b, c, d, e ->
+
+        match a with
+        | "mount" -> AddMount (Mount.MountPoint b, Mount.BdevPath c, Mount.FsType d, Mount.MountOpts e)
+        | "umount" -> RemoveMount (Mount.MountPoint b, Mount.BdevPath c, Mount.FsType d, Mount.MountOpts e)
+        // | "remount" -> toCommand Remount m
```
Comment (johnsonw) [inline IML.Listeners/MountEmitter/src/Transform.fs]: Any reason these are commented out?

File: IML.Types/src/CommandTypes.fs [hunk 14]
```diff
@@ -1,4 +1,3 @@
-// Copyright (c) 2018 Intel Corporation. All rights reserved.
 // Use of this source code is governed by a MIT-style
```
Comment (johnsonw) [inline IML.Types/src/CommandTypes.fs]: Why was the first line removed?

File: IML.Types/src/CommandTypes.fs [hunk 15]
```diff
@@ -213,10 +212,68 @@ module UdevCommand =
...
+  type BdevPath = BdevPath of string
+  [<Erase>]
+  type FsType = FsType of string
+  [<Erase>]
+  type MountOpts = MountOpts of string
+
+type MountCommand =
+  | AddMount of Mount.MountPoint * Mount.BdevPath * Mount.FsType * Mount.MountOpts
+  | RemoveMount of Mount.MountPoint * Mount.BdevPath * Mount.FsType * Mount.MountOpts
+  // | ReplaceMount of Mount.MountPoint * Mount.BdevPath * Mount.FsType * Mount.MountOpts
```
Comment (johnsonw) [inline IML.Types/src/CommandTypes.fs]: Why is this commented out?

File: IML.DeviceScannerDaemon/systemd-units/device-scanner.service (hunk 12 above)
Comment (jgrund) [reply IML.DeviceScannerDaemon/systemd-units/device-scanner.service]: > requires ensures that the service is started without having to do it explicitly

Yes, but given these are collaborator services, I don't think this is the right place to do this.

> after makes sure that the scanner is only started after Mount-emitter

We don't want to pend the device scanner starting on the mount-emitter starting. The device-scanner needs to be the first thing to start.

File: device-scanner.spec [hunk 16]
```diff
@@ -1,17 +1,19 @@
...
-Name:       %{prefix_name}
-Version:    2.1.0
-Release:    2%{?dist}
+%define     proxy_name scanner-proxy
+%define     mount_name mount-emitter
+%define     base_prefixed iml-%{base_name}
+%define     proxy_prefixed iml-%{proxy_name}
+%define     mount_prefixed iml-%{mount_name}
+Name:       %{base_prefixed}
+Version:    2.1.1
```
Comment (jgrund) [inline device-scanner.spec]: Do not bump the version when we haven't released.

File: device-scanner.spec [hunk 17]
```diff
@@ -27,16 +29,16 @@ Requires: iml-node-libzfs
 Requires: socat
 
 %description
-device-scanner-daemon builds an in-memory representation of
-devices using udev and zed.
+%{base-name}-daemon builds an in-memory representation of
```
Comment (jgrund) [inline device-scanner.spec]: This seems like trying to DRY up the code just based on string match, ambivalent on this at the moment, but not sure this is really necessary.

File: IML.DeviceScannerDaemon/systemd-units/device-scanner.service (hunk 12 above)
Comment (tanabarr) [reply IML.DeviceScannerDaemon/systemd-units/device-scanner.service]: so instead of this we should manually start and enable Mount-emitter in spec instead

File: IML.Listeners/MountEmitter/src/Transform.fs [hunk 18]
```diff
@@ -0,0 +1,41 @@
...
+let private notListHeader x = x <> listHeader
+
+let transform (x:Stream.Readable<string>) =
+  x
+    |> Stream.map (buffer.Buffer.from >> Ok)
+    |> Stream.LineDelimited.create()
+    |> Stream.map (fun x -> Ok (x.Split([| ' ' |], System.StringSplitOptions.RemoveEmptyEntries)))
+    |> Stream.map toRow
+    |> Stream.filter (notPollHeader >> Ok)
+    |> Stream.filter (notListHeader >> Ok)
```
Comment (jgrund) [inline IML.Listeners/MountEmitter/src/Transform.fs]: Could probably do a single pattern match since the types are similar:

```
function
   | (_, "TARGET", "SOURCE", "FSTYPE", "OPTIONS") -> true
   | _ -> false
```

File: IML.Listeners/MountEmitter/src/Transform.fs [hunk 19]
```diff
@@ -8,15 +8,17 @@ open Fable.Import.Node
...
 
-type Row = (string *  string * string * string * string)
+type Row = (string * string * string * string * string * string * string)
 
 let private toRow = function
-  | [| a; b; c; d; e; |] -> Ok (a, b, c, d, e)
-  | [| a; b; c; d; |] -> Ok ("mount", a, b, c, d)
+  // findmnt --poll -o ACTION,TARGET,SOURCE,FSTYPE,OPTIONS,OLD-TARGET,OLD-OPTIONS
+  | [| a; b; c; d; e; f; g; |] -> Ok (a, b, c, d, e, f, g)
+  // findmnt --list
```
Comment (jgrund) [inline IML.Listeners/MountEmitter/src/Transform.fs]: Does --list have a -o flag?

File: IML.Listeners/MountEmitter/src/Transform.fs [hunk 20]
```diff
@@ -0,0 +1,56 @@
...
+    |> Stream.map(function
+      | a, b, c, d, e, f, g ->
+        let short =
+          (
+            Mount.MountPoint b,
+            Mount.BdevPath c,
+            Mount.FsType d,
+            Mount.MountOpts e
+          )
+        let long =
```
Comment (jgrund) [inline IML.Listeners/MountEmitter/src/Transform.fs]: Seems like you could always define `long`, and construct `short` from it by deconstructing the first four items into it's own tuple

File: IML.Listeners/MountEmitter/src/Transform.fs [hunk 21]
```diff
@@ -0,0 +1,56 @@
...
+            Mount.MountOpts e,
+            Mount.MountPoint f,
+            Mount.MountOpts g
+          )
+        match a with
+        | "mount" -> AddMount short
+        | "umount" -> RemoveMount short
+        | "remount" -> ReplaceMount long
+        | "move" -> MoveMount long
+        | _ -> failwithf "did not get expected row, got %A" a
```
Comment (jgrund) [inline IML.Listeners/MountEmitter/src/Transform.fs]: Instead of failwithf, return an `Error` here

File: IML.Listeners/MountEmitter/test/TransformTest.fs [hunk 22]
```diff
@@ -0,0 +1,62 @@
+module IML.MountEmitter.TransformTest
+
+open Fable.Import.Jest
+open Fable.Import.Node.PowerPack.Stream
+open Transform
+open Fable.PowerPack
+
+let streamTap =
+  transform
+    >> tap (fun xs -> expect.Invoke(xs).toMatchSnapshot())
```
Comment (jgrund) [inline IML.Listeners/MountEmitter/test/TransformTest.fs]: Probably better to do the match afterwards, I.E. always use `promiseMatch`.

the `tap` will be called for every chunk through the stream.

File: IML.Listeners/MountEmitter/test/TransformTest.fs [hunk 23]
```diff
@@ -0,0 +1,62 @@
...
+    >> Util.streamToPromise
+
+let promiseMatch =
+  transform
+    >> Util.streamToPromise
+    >> Promise.map (fun xs -> expect.Invoke(xs).toMatchSnapshot())
```
Comment (jgrund) [inline IML.Listeners/MountEmitter/test/TransformTest.fs]: Can be shortened to:

`Promise.map toMatchSnapshot`

Comment (jgrund) [inline IML.Listeners/MountEmitter/test/TransformTest.fs]: You probably want to encode the output before taking a snapshot. The output will be more readable as JSON than the fable representation.

File: IML.Types/test/CommandTypesTest.fs [hunk 24]
```diff
@@ -115,6 +137,20 @@ udevCommands
...
+
+    let o:string =
+      !!snaps?(caseName)
+      |> String.filter (fun x -> x <> '\n')
+
+    Test(name, fun () ->
+      mountDecode !!(JS.JSON.parse o) == Ok cmd
+    )
+  )
+  |> testList "decoding Mounnt commands"
```
Comment (jgrund) [inline IML.Types/test/CommandTypesTest.fs]: extra n in Mount

File: device-scanner.sln [hunk 25]
```diff
@@ -51,11 +51,15 @@ Project("{F2A71F9B-5D33-465A-A702-920D77279786}") = "IML.Types", "IML.Types\src\
...
 EndProject
-Project("{F2A71F9B-5D33-465A-A702-920D77279786}") = "IML.CommonLibrary", "IML.CommonLibrary\src\IML.CommonLibrary.fsproj", "{4FE85420-3C4B-459B-AF38-68C3D717A0EB}"
+Project("{F2A71F9B-5D33-465A-A702-920D77279786}") = "MountEmitter", "IML.Listeners\MountEmitter\src\MountEmitter.fsproj", "{7F2E0C7B-6DCC-496F-8ABE-DD0E6EE860A9}"
 EndProject
-Project("{F2A71F9B-5D33-465A-A702-920D77279786}") = "IML.CommonLibraryTest", "IML.CommonLibrary\test\IML.CommonLibraryTest.fsproj", "{B5BDA837-2D74-4D1F-8BC1-F9CFE4569B8B}"
+Project("{F2A71F9B-5D33-465A-A702-920D77279786}") = "MountEmitterTest", "IML.Listeners\MountEmitter\test\MountEmitterTest.fsproj", "{F7C35BDF-6FBB-4056-8228-018FCAD53DE1}"
+EndProject
+Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "IML.CommonLibrary", "IML.CommonLibrary", "{BC78157D-9A4C-4DC2-BEF5-25A1E4F40D70}"
+EndProject
+Project("{F2A71F9B-5D33-465A-A702-920D77279786}") = "IML.CommonLibraryTest", "IML.CommonLibrary\test\IML.CommonLibraryTest.fsproj", "{4D242930-AC48-4C59-8301-04C83D8CDBEA}"
```
Comment (jgrund) [inline device-scanner.sln]: It seems like the `CommonLibrary.fsproj` was lost in these changes. Here is output of `dotnet sln list`:

```
IML.DeviceScannerDaemon
IML.DeviceScannerDaemon/src/IML.DeviceScannerDaemon.fsproj
IML.DeviceScannerDaemon/test/IML.DeviceScannerDaemonTest.fsproj
IML.Listeners
IML.Listeners/PoolImportZedlet/PoolImportZedlet.fsproj
IML.Listeners/PoolExportZedlet/PoolExportZedlet.fsproj
IML.Listeners/PoolDestroyZedlet/PoolDestroyZedlet.fsproj
IML.Listeners/PoolCreateZedlet/PoolCreateZedlet.fsproj
IML.Listeners/HistoryEventZedlet/HistoryEventZedlet.fsproj
IML.Listeners/VdevAddZedlet/VdevAddZedlet.fsproj
IML.Listeners/UdevListener/UdevListener.fsproj
IML.StatefulPromise
IML.StatefulPromise/source/IML.StatefulPromise.fsproj
IML.StatefulPromise/test/IML.StatefulPromiseTest.fsproj
IML.IntegrationTestFramework
IML.IntegrationTestFramework/source/IML.IntegrationTestFramework.fsproj
IML.IntegrationTestFramework/test/IML.IntegrationTestFrameworkTest.fsproj
IML.IntegrationTest/IML.IntegrationTest.fsproj
IML.ScannerProxyDaemon
IML.ScannerProxyDaemon/src/IML.ScannerProxyDaemon.fsproj
IML.ScannerProxyDaemon/test/IML.ScannerProxyDaemonTest.fsproj
IML.Types
IML.Types/src/IML.Types.fsproj
IML.Types/test/IML.TypesTest.fsproj
MountEmitter
IML.Listeners/MountEmitter/src/MountEmitter.fsproj
IML.Listeners/MountEmitter/test/MountEmitterTest.fsproj
IML.CommonLibrary
IML.CommonLibrary/test/IML.CommonLibraryTest.fsproj
```

File: IML.Listeners/MountEmitter/src/Transform.fs (hunk 19 above)
Comment (tanabarr) [reply IML.Listeners/MountEmitter/src/Transform.fs]: yes but I'm not sure how that would quite help, some of the columns are not available without --poll

File: device-scanner.sln (hunk 25 above)
Comment (tanabarr) [reply device-scanner.sln]: this file has been updated

File: IML.Listeners/MountEmitter/src/MountEmitter.fsproj [hunk 26]
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (jgrund) [inline IML.Listeners/MountEmitter/src/MountEmitter.fsproj]: This file should be called `IML.MountEmitter.fsproj`.

Comment (tanabarr) [reply IML.Listeners/MountEmitter/src/MountEmitter.fsproj]: ah, was trying to work that one out, thanks

File: IML.Listeners/MountEmitter/src/Transform.fs [hunk 27]
```diff
@@ -0,0 +1,76 @@
...
+let private toRow = function
+  // findmnt --list (record or header)
+  | [| a; b; c; d; |] -> Ok ("mount", a, b, c, d, "", "")
+  // mount
+  | [| a; b; c; d; e; |] -> Ok (a, b, c, d, e, "", "")
+  // remount or move
+  | [| a; b; c; d; e; f; |] -> Ok (a, b, c, d, e, f, "")
+  // umount or header
+  | [| a; b; c; d; e; f; g; |] -> Ok (a, b, c, d, e, f, g)
+  | x -> failwithf "did not get expected row data, got %A" x
```
Comment (jgrund) [inline IML.Listeners/MountEmitter/src/Transform.fs]:
```f#
| x -> Error (exn (sprintf "did not get expected row data, got %A" x))
```

File: IML.Listeners/MountEmitter/src/Transform.fs [hunk 28]
```diff
@@ -0,0 +1,76 @@
...
+
+let transform (x:Stream.Readable<string>) =
+  x
+    |> Stream.map (buffer.Buffer.from >> Ok)
+    |> Stream.LineDelimited.create()
+    |> Stream.map (fun x -> Ok (x.Split([| ' ' |], System.StringSplitOptions.RemoveEmptyEntries)))
+    |> Stream.map toRow
+    |> Stream.filter (notHeader >> Ok)
+    |> Stream.map(function
+      | a, b, c, d, e, f, _ ->
```
Comment (jgrund) [inline IML.Listeners/MountEmitter/src/Transform.fs]: If we don't use `g`, do we need it?

File: IML.Listeners/MountEmitter/src/Transform.fs [hunk 29]
```diff
@@ -0,0 +1,76 @@
...
+          RemoveMount short
+            |> Command.MountCommand |> Ok
+        | "remount" ->
+          ReplaceMount
+            (
+              Mount.MountPoint b,
+              Mount.BdevPath c,
+              Mount.FsType d,
+              Mount.MountOpts e,
+              Mount.MountOpts f
```
Comment (jgrund) [inline IML.Listeners/MountEmitter/src/Transform.fs]: Is this old-target?

File: IML.Listeners/MountEmitter/src/Transform.fs (hunk 28 above)
Comment (tanabarr) [reply IML.Listeners/MountEmitter/src/Transform.fs]: need to match against it for header

File: IML.Listeners/MountEmitter/src/Transform.fs (hunk 29 above)
Comment (tanabarr) [reply IML.Listeners/MountEmitter/src/Transform.fs]: either that (moveMount) or old-opts (reMount)

File: IML.Listeners/MountEmitter/src/Transform.fs (hunk 28 above)
Comment (jgrund) [reply IML.Listeners/MountEmitter/src/Transform.fs]: Couldn't header filter be:


```f#
let private notHeader = function
  | (_, "TARGET", "SOURCE", "FSTYPE", "OPTIONS", _) -> false
  | _ -> true
```
?

File: IML.Listeners/MountEmitter/src/Transform.fs (hunk 29 above)
Comment (jgrund) [reply IML.Listeners/MountEmitter/src/Transform.fs]: So columns are variant based on action?

Comment (jgrund) [reply IML.Listeners/MountEmitter/src/Transform.fs]: Ah, sort of.

It's all dependent on column widths

File: IML.Listeners/MountEmitter/src/Transform.fs [hunk 30]
```diff
@@ -0,0 +1,79 @@
...
+// license that can be found in the LICENSE file.
+
+module IML.MountEmitter.Transform
+
+open Fable.Import.Node
+open Fable.Import.Node.PowerPack
+open IML.Types.CommandTypes
+
+// findmnt --poll -o ACTION,TARGET,SOURCE,FSTYPE,OPTIONS,OLD-TARGET,OLD-OPTIONS
+type Row = (string * string * string * string * string * string * string)
```
Comment (jgrund) [inline IML.Listeners/MountEmitter/src/Transform.fs]: Is this type referenced anywhere?

File: IML.Listeners/MountEmitter/src/Transform.fs (hunk 29 above)
Comment (tanabarr) [reply IML.Listeners/MountEmitter/src/Transform.fs]: old-opts value not present in move action for example

File: IML.Listeners/MountEmitter/test/TransformTest.fs [hunk 31]
```diff
@@ -0,0 +1,106 @@
...
+open Fable.Import.Node.PowerPack.Stream
+open Transform
+open IML.Types.CommandTypes
+open Fable.PowerPack
+open Matchers
+
+let promiseMatch =
+  transform
+    >> Util.streamToPromise
+    >> Promise.map ((List.map Command.encoder) >> toMatchSnapshot)
```
Comment (jgrund) [inline IML.Listeners/MountEmitter/test/TransformTest.fs]: Convert this to an array, the snapshot will be cleaner.

File: IML.Listeners/MountEmitter/systemd-units/mount-emitter.service [hunk 32]
```diff
@@ -0,0 +1,13 @@
+[Unit]
+Description=IML Mount Emitter
+
+[Service]
+Restart=always
+Environment=NODE_ENV=production
+ExecStartPre=stdbuf -o L findmnt --list | /usr/lib64/iml-mount-emitter/mount-emitter
```
Comment (jgrund) [inline IML.Listeners/MountEmitter/systemd-units/mount-emitter.service]: If we use the `-n` flag, we can drop the header parsing.

File: IML.Listeners/MountEmitter/systemd-units/mount-emitter.service [hunk 33]
```diff
@@ -0,0 +1,13 @@
...
+ExecStart=stdbuf -o L findmnt --poll -o ACTION,TARGET,SOURCE,FSTYPE,OPTIONS,OLD-TARGET,OLD-OPTIONS | /usr/lib64/iml-mount-emitter/mount-emitter
```
Comment (jgrund) [inline IML.Listeners/MountEmitter/systemd-units/mount-emitter.service]: If we use the `-n` flag, we can drop the header parsing.

File: IML.Listeners/MountEmitter/systemd-units/mount-emitter.service (hunk 32 above)
Comment (jgrund) [reply IML.Listeners/MountEmitter/systemd-units/mount-emitter.service]: Or, even better we could use the `-P` flag:

```
-P, --pairs
              Use key="value" output format.  All potentially unsafe characters are hex-escaped (\x<code>).
```

File: IML.Listeners/MountEmitter/src/Transform.fs [hunk 34]
```diff
@@ -0,0 +1,80 @@
...
+let private toError =
+  exn >> Error
+
+let private columnError = function
+  | x -> toError (sprintf "did not find expected column values for '%A' action" x)
+
+let private toSuccess =
+  Command.MountCommand >> Ok
+
+let private toMap (xs:string []) =
```
Comment (jgrund) [inline IML.Listeners/MountEmitter/src/Transform.fs]: Here's a way to do this without the mutable Map and the nested reduce:

```f#
let toMap =
  Array.fold (fun acc (x:string) ->
    let idx = x.IndexOf "="
    let k = x.Substring(0, idx)
    let v = x.Substring(idx + 1).Trim('"')

    Map.add k v acc
  ) Map.empty
  >> Ok
```

File: IML.Listeners/MountEmitter/src/Transform.fs [hunk 35]
```diff
@@ -0,0 +1,80 @@
...
+    |> Stream.map (fun x -> Ok (x.Split([| ' ' |], System.StringSplitOptions.RemoveEmptyEntries)))
+    |> Stream.map toMap
+    |> Stream.map (fun m ->
+      let short =
+        (
+          Mount.MountPoint (Option.get (m.TryFind "TARGET")),
+          Mount.BdevPath (Option.get (m.TryFind "SOURCE")),
+          Mount.FsType (Option.get (m.TryFind "FSTYPE")),
+          Mount.MountOpts (Option.get (m.TryFind "OPTIONS"))
+        )
```
Comment (jgrund) [inline IML.Listeners/MountEmitter/src/Transform.fs]: Don't need to unwrap an `Option` if you use `Map.find`:

```f#
      let short =
        (
          Mount.MountPoint (Map.find "TARGET" m),
          Mount.BdevPath (Map.find "SOURCE" m),
          Mount.FsType (Map.find "FSTYPE" m),
          Mount.MountOpts (Map.find "OPTIONS" m)
        )
```

File: IML.Listeners/MountEmitter/src/Transform.fs [hunk 36]
```diff
@@ -0,0 +1,80 @@
...
+            MoveMount
+              (
+                target, source, fstype, opts,
+                Mount.MountPoint y
+              ) |> toSuccess
+          | None -> columnError x
+        | _ ->
+          toError (sprintf "unexpected action type, received %A" x)
+      // no ACTION key is populated when --poll option is not used
+      | None -> AddMount short |> toSuccess
```
Comment (jgrund) [inline IML.Listeners/MountEmitter/src/Transform.fs]: This can be simplified to remove the nested match statements:

```f#
      let short =
        (
          Mount.MountPoint (Map.find "TARGET" m),
          Mount.BdevPath (Map.find "SOURCE" m),
          Mount.FsType (Map.find "FSTYPE" m),
          Mount.MountOpts (Map.find "OPTIONS" m)
        )
      let (target, source, fstype, opts) = short
      match m.TryFind "ACTION" with
        | Some "mount" ->
          AddMount short |> toSuccess
        | Some "umount" ->
          RemoveMount short |> toSuccess
        | Some "remount" ->
          ReplaceMount (
            target, source, fstype, opts,
            Mount.MountOpts (Map.find "OLD-TARGET" m)
          )
            |> toSuccess
        | Some "move" ->
          MoveMount(
            target, source, fstype, opts,
            Mount.MountPoint (Map.find "OLD-TARGET" m)
          )
            |> toSuccess
        // no ACTION key is populated when --poll option is not used
        | None -> AddMount short |> toSuccess
        | _ ->
          toError (sprintf "unexpected action type, received %A" x)
```

Comment (tanabarr) [reply IML.Listeners/MountEmitter/src/Transform.fs]: interesting that you remove the exception handling in Replace and Move actions, I had put that in because I thought you would asked me to put it in if I didn't

Comment (jgrund) [reply IML.Listeners/MountEmitter/src/Transform.fs]: Given it's going to raise in either case I think it should be ok.

File: IML.Listeners/MountEmitter/systemd-units/mount-emitter.service [hunk 37]
```diff
@@ -4,8 +4,8 @@ Description=IML Mount Emitter
 [Service]
 Restart=always
 Environment=NODE_ENV=production
-ExecStartPre=stdbuf -o L findmnt -P | /usr/lib64/iml-mount-emitter/mount-emitter
-ExecStart=stdbuf -o L findmnt --poll -P -o ACTION,TARGET,SOURCE,FSTYPE,OPTIONS,OLD-TARGET,OLD-OPTIONS | /usr/lib64/iml-mount-emitter/mount-emitter
+ExecStartPre=/usr/bin/sh -c "/usr/bin/findmnt -P | /usr/lib64/iml-mount-emitter/mount-emitter"
```
Comment (jgrund) [inline IML.Listeners/MountEmitter/systemd-units/mount-emitter.service]: I think we want to use exec for these

**Comment by jgrund** [comment]
Nice Work!
